Master branch (date=20210413): flag --cluster.listen-address= or --cluster.listen-address="" produces segmentation violation.

<!--

    Please do *NOT* ask usage questions in Github issues.

    If your issue is not a feature request or bug report use:
    https://groups.google.com/forum/#!forum/prometheus-users. If
    you are unsure whether you hit a bug, search and ask in the
    mailing list first.

    You can find more information at: https://prometheus.io/community/

-->

**What did you do?**

Since today, I discovered the tag _master_, 🥰  on docker and I changed the tags for `blackbox_exporter` ✔️  and `alertmanager` ❎ . It makes sense since the latest tagged versions are "pretty" old. Changing the tag to `master` is a good recipe, well, maybe sometimes not so good ...

While changing the docker tag from _latest_ (`date=20200617-08:54:02`) to _master_ (`date=20210413-15:17:06`) on `alertmanager`, appeared the failure. 

**What did you expect to see?**

A beautiful 💚 world 

**What did you see instead? Under which circumstances?**

Segmentation violation on `alartmanager` logs, `502` on GUI (status) 💔 

Appx every 15 seconds the alertmanager container dies, starts again and fails again cyclically.

**Environment**

* System information:

`Linux 5.11.0-7612-generic x86_64`

* Alertmanager version:

`version=0.21.0, branch=master, revision=22ac6dff21901bfce14545da59b37a1aaca0db3a`

* Prometheus version:

`version=2.26.0, branch=HEAD, revision=3cafc58827d1ebd1a67749f88be4218f0bab3d8d`

* Alertmanager configuration file:
```
not relevant, see below
```

* Prometheus configuration file:
```
not relevant, see below
```

* docker run:

```
docker run -d -p 9093:9093 \
     --restart=unless-stopped \
     --hostname=alertmanager \
     --name alertmanager \
     -v /usr/local/etc/alertmanager:/etc/alertmanager \
     -v /etc/timezone:/etc/timezone:ro \
     -v /etc/localtime:/etc/localtime:ro \
     -e TZ=Europe/Berlin \
     prom/alertmanager:master \
     --config.file="/etc/alertmanager/alertmanager.yml" \
     --storage.path=/etc/alertmanager/data \
     --data.retention=168h \
     --web.external-url=https://alertmanager.domain.xy \
     --log.level=debug \
     --cluster.listen-address=
```


* Logs:
```
level=info ts=2021-04-17T16:10:34.766Z caller=main.go:221 msg="Starting Alertmanager" version="(version=0.21.0, branch=master, revision=22ac6dff21901bfce14545da59b37a1aaca0db3a)"
level=info ts=2021-04-17T16:10:34.766Z caller=main.go:222 build_context="(go=go1.16.3, user=root@c386f44bafa0, date=20210413-15:17:06)"
level=debug ts=2021-04-17T16:10:34.827Z caller=main.go:360 externalURL=https://alertmanager.domain.xy
level=info ts=2021-04-17T16:10:34.827Z caller=coordinator.go:113 component=configuration msg="Loading configuration file" file=/etc/alertmanager/alertmanager.yml
level=info ts=2021-04-17T16:10:34.828Z caller=coordinator.go:126 component=configuration msg="Completed loading of configuration file" file=/etc/alertmanager/alertmanager.yml
level=debug ts=2021-04-17T16:10:34.834Z caller=main.go:477 routePrefix=/
level=info ts=2021-04-17T16:10:34.835Z caller=main.go:497 msg=Listening address=:9093
level=info ts=2021-04-17T16:10:34.838Z caller=tls_config.go:191 msg="TLS is disabled." http2=false
level=debug ts=2021-04-17T16:11:17.371Z caller=dispatch.go:138 component=dispatcher msg="Received alert" alert="Let's Encrypt ssl certificate will expire within 1 Month[36fbef0][resolved]"
level=debug ts=2021-04-17T16:11:17.371Z caller=dispatch.go:474 component=dispatcher aggrGroup="{}:{alertname=\"Let's Encrypt ssl certificate will expire within 1 Month\", server=\"arkanoid\"}" msg=flushing alerts="[Let's Encrypt ssl certificate will expire within 1 Month[36fbef0][resolved]]"
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0xafe865]
goroutine 152 [running]:
github.com/prometheus/alertmanager/cluster.(*Peer).WaitReady(0x0, 0x113be38, 0xc00017c210, 0x125, 0x170)
        /app/cluster/cluster.go:591 +0x45
github.com/prometheus/alertmanager/notify.(*GossipSettleStage).Exec(0xc0008074e0, 0x113be38, 0xc00017c210, 0x11258a0, 0xc00036a090, 0xc0002884e0, 0x1, 0x1, 0x1111fb0, 0x1111fb0, ...)
        /app/notify/notify.go:433 +0xa3
github.com/prometheus/alertmanager/notify.MultiStage.Exec(0xc00062afa0, 0x5, 0x5, 0x113be38, 0xc00017c210, 0x11258a0, 0xc00036a090, 0xc0002884e0, 0x1, 0x1, ...)
        /app/notify/notify.go:385 +0x102
github.com/prometheus/alertmanager/notify.RoutingStage.Exec(0xc00036a000, 0x113be38, 0xc00017c210, 0x11258a0, 0xc00036a090, 0xc0002884e0, 0x1, 0x1, 0x14, 0x0, ...)
        /app/notify/notify.go:371 +0x1e2
github.com/prometheus/alertmanager/dispatch.(*Dispatcher).processAlert.func1(0x113be38, 0xc00017c210, 0xc0002884e0, 0x1, 0x1, 0xc000124140)
        /app/dispatch/dispatch.go:300 +0xa6
github.com/prometheus/alertmanager/dispatch.(*aggrGroup).run.func1(0xc0002884e0, 0x1, 0x1, 0x4)
        /app/dispatch/dispatch.go:416 +0x55
github.com/prometheus/alertmanager/dispatch.(*aggrGroup).flush(0xc00010b7a0, 0xc0002eff60)
        /app/dispatch/dispatch.go:476 +0x529
github.com/prometheus/alertmanager/dispatch.(*aggrGroup).run(0xc00010b7a0, 0xc00019ce00)
        /app/dispatch/dispatch.go:415 +0x5a5
created by github.com/prometheus/alertmanager/dispatch.(*Dispatcher).processAlert
        /app/dispatch/dispatch.go:299 +0x279

```

### Is the error reproduceable? if yes, how?
Definitely yes. Triggering the segmentation violation is reproduceable and related to the alertmanager flag `--cluster.listen-address=` or `--cluster.listen-address=""`. 

### Temporary solution
As workaround, I omited the `cluster.*` flag. It looks trivial, but it took a time to analyze and to came up to the conclusion. 

The tag `latest` on docker hasn't the bug and works as expected 🚀

Same in latest 0.22.0-rc0 version:

```
level=info ts=2021-05-06T07:21:24.029Z caller=main.go:221 msg="Starting Alertmanager" version="(version=0.22.0-rc.0, branch=HEAD, revision=8ec96dbfeb3cbe5d160e20dd57520996cd2633e6)"
level=info ts=2021-05-06T07:21:24.029Z caller=main.go:222 build_context="(go=go1.16.3, user=root@4bf45e384ce7, date=20210504-19:46:48)"
level=info ts=2021-05-06T07:21:24.066Z caller=coordinator.go:113 component=configuration msg="Loading configuration file" file=/opt/prometheus/etc/alertmanager/am.yml
level=info ts=2021-05-06T07:21:24.068Z caller=coordinator.go:126 component=configuration msg="Completed loading of configuration file" file=/opt/prometheus/etc/alertmanager/am.yml
level=info ts=2021-05-06T07:21:24.076Z caller=main.go:497 msg=Listening address=:9093
level=info ts=2021-05-06T07:21:24.076Z caller=tls_config.go:191 msg="TLS is disabled." http2=false
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x50 pc=0xb06f45]
goroutine 154 [running]:
github.com/prometheus/alertmanager/cluster.(*Peer).WaitReady(0x0, 0x115c038, 0xc000978330, 0x3e, 0x40db1b)
/app/cluster/cluster.go:594 +0x45
github.com/prometheus/alertmanager/notify.(*GossipSettleStage).Exec(0xc000306910, 0x115c038, 0xc000978330, 0x1145920, 0xc000937f80, 0xc00000ed08, 0x1, 0x1, 0x1131ae8, 0x1131ae8, ...)
/app/notify/notify.go:433 +0xa3
github.com/prometheus/alertmanager/notify.MultiStage.Exec(0xc0005f4eb0, 0x5, 0x5, 0x115c038, 0xc000978330, 0x1145920, 0xc000937f80, 0xc00000ed08, 0x1, 0x1, ...)
/app/notify/notify.go:385 +0x102
github.com/prometheus/alertmanager/notify.RoutingStage.Exec(0xc000937410, 0x115c038, 0xc000978330, 0x1145920, 0xc000937f80, 0xc00000ed08, 0x1, 0x1, 0xf, 0x19, ...)
/app/notify/notify.go:371 +0x1e2
github.com/prometheus/alertmanager/dispatch.(*Dispatcher).processAlert.func1(0x115c038, 0xc000978330, 0xc00000ed08, 0x1, 0x1, 0xc0000be000)
/app/dispatch/dispatch.go:301 +0xa6
github.com/prometheus/alertmanager/dispatch.(*aggrGroup).run.func1(0xc00000ed08, 0x1, 0x1, 0x4)
/app/dispatch/dispatch.go:417 +0x55
github.com/prometheus/alertmanager/dispatch.(*aggrGroup).flush(0xc000614cf0, 0xc000965f60)
/app/dispatch/dispatch.go:477 +0x529
github.com/prometheus/alertmanager/dispatch.(*aggrGroup).run(0xc000614cf0, 0xc0003046e0)
/app/dispatch/dispatch.go:416 +0x5a5
created by github.com/prometheus/alertmanager/dispatch.(*Dispatcher).processAlert
/app/dispatch/dispatch.go:300 +0x279

```

--- a/cmd/alertmanager/main.go
+++ b/cmd/alertmanager/main.go
@@ -426,6 +426,15 @@ func run() int {
 
 		inhibitor = inhibit.NewInhibitor(alerts, conf.InhibitRules, marker, logger)
 		silencer := silence.NewSilencer(silences, marker, logger)
+
+		// An interface value that holds a nil concrete value is non-nil.
+		// Therefore we explicly pass an empty interface, to detect if the
+		// cluster is not enabled in notify.
+		var pipelinePeer notify.Peer
+		if peer != nil {
+			pipelinePeer = peer
+		}
+
 		pipeline := pipelineBuilder.New(
 			receivers,
 			waitFunc,
@@ -433,7 +442,7 @@ func run() int {
 			silencer,
 			muteTimes,
 			notificationLog,
-			peer,
+			pipelinePeer,
 		)
 		configuredReceivers.Set(float64(len(activeReceivers)))
 		configuredIntegrations.Set(float64(integrationsNum))

